File: verify_numbers.mjs
import Database from 'duckdb';
import fetch from 'node-fetch';

const DB_PATH = 'D:/darshanmeel/AURA/data/aura_read.duckdb';
const API_URL = 'http://localhost:3000/api/dashboard?range=all';

let db;
let conn;

async function initDB() {
  db = new Database(DB_PATH);
  conn = await db.connect();
}

async function runQuery(sql) {
  const result = await conn.all(sql);
  return result;
}

async function fetchAPI() {
  const res = await fetch(API_URL);
  if (!res.ok) throw new Error(`API failed: ${res.status}`);
  return await res.json();
}

function compareValues(label, apiValue, dbValue, tolerance = 0.01) {
  if (apiValue === null && dbValue === null) {
    console.log(`${label.padEnd(35)} api=null        db=null        MATCH`);
    return true;
  }

  // Handle numeric comparisons with tolerance for floats
  if (typeof apiValue === 'number' && typeof dbValue === 'number') {
    const delta = Math.abs(apiValue - dbValue);
    const relDelta = Math.abs((apiValue - dbValue) / (dbValue || 1));
    if (delta < 0.01 || relDelta < tolerance) {
      console.log(`${label.padEnd(35)} api=${apiValue}        db=${dbValue}        MATCH`);
      return true;
    }
  }

  // Handle string comparisons
  if (apiValue === dbValue) {
    console.log(`${label.padEnd(35)} api=${apiValue}        db=${dbValue}        MATCH`);
    return true;
  }

  console.log(
    `${label.padEnd(35)} api=${apiValue}        db=${dbValue}        MISMATCH (delta=${
      typeof apiValue === 'number' && typeof dbValue === 'number' ? apiValue - dbValue : 'N/A'
    })`
  );
  return false;
}

async function main() {
  await initDB();
  const api = await fetchAPI();

  console.log('Verifying dashboard numbers...\n');

  let matches = 0;
  let mismatches = 0;

  // KPIs
  if (api.kpis) {
    const kpi = api.kpis;
    console.log('=== KPIs ===');

    const sessionsResult = await runQuery(`SELECT COUNT(DISTINCT session_id) AS cnt FROM dim_sessions`);
    if (compareValues('total_sessions', kpi.total_sessions, sessionsResult[0]?.cnt)) matches++;
    else mismatches++;

    const costResult = await runQuery(`SELECT SUM(total_cost) AS cost FROM dim_sessions`);
    if (compareValues('total_cost', kpi.total_cost, costResult[0]?.cost)) matches++;
    else mismatches++;

    const turnsResult = await runQuery(`SELECT SUM(turn_count) AS turns FROM dim_sessions`);
    if (compareValues('total_turns', kpi.total_turns, turnsResult[0]?.turns)) matches++;
    else mismatches++;

    const toolsResult = await runQuery(`SELECT SUM(tools_used) AS tools FROM dim_sessions`);
    if (compareValues('total_tool_calls', kpi.total_tool_calls, toolsResult[0]?.tools)) matches++;
    else mismatches++;

    const commitsResult = await runQuery(`SELECT SUM(commits) AS commits FROM dim_sessions`);
    if (compareValues('total_commits', kpi.total_commits, commitsResult[0]?.commits)) matches++;
    else mismatches++;

    const appsResult = await runQuery(`SELECT COUNT(DISTINCT cwd) AS apps FROM dim_sessions`);
    if (compareValues('total_apps', kpi.total_apps, appsResult[0]?.apps)) matches++;
    else mismatches++;

    const peopleResult = await runQuery(`SELECT COUNT(DISTINCT person_id) AS people FROM dim_sessions`);
    if (compareValues('total_people', kpi.total_people, peopleResult[0]?.people)) matches++;
    else mismatches++;

    const activeResult = await runQuery(
      `SELECT COUNT(DISTINCT CASE WHEN status = 'active' THEN session_id END) AS active FROM dim_sessions`
    );
    if (compareValues('active_sessions', kpi.active_sessions, activeResult[0]?.active)) matches++;
    else mismatches++;

    const cacheResult = await runQuery(`
      SELECT
        SUM(cache_read_total) / NULLIF(SUM(cache_read_total + ephemeral_5m_total + ephemeral_1h_total), 0) AS rate
      FROM dim_sessions
    `);
    if (compareValues('cache_hit_rate', kpi.cache_hit_rate, cacheResult[0]?.rate, 0.001)) matches++;
    else mismatches++;

    console.log();
  }

  // Daily Spend
  if (api.dailySpend && api.dailySpend.length > 0) {
    console.log('=== Daily Spend (sample) ===');
    const firstDay = api.dailySpend[0];
    const dailyResult = await runQuery(`SELECT SUM(daily_cost) AS cost FROM fact_daily_spend WHERE date = '${firstDay.date}'`);
    if (compareValues('daily_cost_sample', firstDay.cost, dailyResult[0]?.cost)) matches++;
    else mismatches++;
    console.log();
  }

  // Top Apps
  if (api.topApps && api.topApps.length > 0) {
    console.log('=== Top Apps (sample) ===');
    if (compareValues('topApps_row_count', api.topApps.length, api.topApps.length)) matches++;
    if (api.topApps[0]) {
      const firstAppResult = await runQuery(`
        SELECT SUM(ds.total_cost) AS total_cost FROM dim_sessions ds
        LEFT JOIN int_app_cwd_lookup al ON al.cwd = ds.cwd AND al.tenant_id = ds.tenant_id
        LEFT JOIN dim_apps da ON da.app_id = al.app_id
        WHERE COALESCE(da.app_id, ds.cwd) = '${api.topApps[0].app_id}'
      `);
      if (compareValues('topApps[0]_cost', api.topApps[0].total_cost, firstAppResult[0]?.total_cost)) matches++;
      else mismatches++;
    }
    console.log();
  }

  // Top Agents
  if (api.topAgents && api.topAgents.length > 0) {
    console.log('=== Top Agents (sample) ===');
    if (api.topAgents[0]) {
      const firstAgentResult = await runQuery(`SELECT SUM(ds.total_cost) AS cost FROM dim_sessions ds WHERE ds.agent = '${api.topAgents[0].agent}'`);
      if (compareValues('topAgents[0]_cost', api.topAgents[0].total_cost, firstAgentResult[0]?.cost)) matches++;
      else mismatches++;
    }
    console.log();
  }

  // Tool Mix
  if (api.toolMix && api.toolMix.length > 0) {
    console.log('=== Tool Mix (sample) ===');
    if (api.toolMix[0]) {
      const firstToolResult = await runQuery(`SELECT COUNT(*) AS cnt FROM fact_tool_executions WHERE tool_name = '${api.toolMix[0].tool_name}'`);
      if (compareValues('toolMix[0]_count', api.toolMix[0].call_count, firstToolResult[0]?.cnt)) matches++;
      else mismatches++;
    }
    console.log();
  }

  // Providers
  if (api.providers && api.providers.length > 0) {
    console.log('=== Providers (sample) ===');
    if (api.providers[0]) {
      const firstProviderResult = await runQuery(`SELECT SUM(total_cost) AS cost FROM dim_sessions WHERE provider = '${api.providers[0].provider}'`);
      if (compareValues('providers[0]_cost', api.providers[0].cost, firstProviderResult[0]?.cost)) matches++;
      else mismatches++;
    }
    console.log();
  }

  // Models
  if (api.models && api.models.length > 0) {
    console.log('=== Models (sample) ===');
    if (api.models[0]) {
      const firstModelResult = await runQuery(`SELECT SUM(total_cost) AS cost FROM dim_sessions WHERE model = '${api.models[0].model}'`);
      if (compareValues('models[0]_cost', api.models[0].cost, firstModelResult[0]?.cost)) matches++;
      else mismatches++;
    }
    console.log();
  }

  // Recent Errors
  if (api.recentErrors !== undefined && api.recentErrors !== null) {
    console.log('=== Recent Errors (row count) ===');
    const errorsResult = await runQuery(`SELECT COUNT(*) AS cnt FROM fact_errors`);
    if (compareValues('recentErrors_exists', api.recentErrors.length > 0, errorsResult[0]?.cnt > 0)) matches++;
    else mismatches++;
    console.log();
  }

  // Top Files
  if (api.topFiles && api.topFiles.length > 0) {
    console.log('=== Top Files (sample) ===');
    if (api.topFiles[0]) {
      const firstFileResult = await runQuery(`SELECT SUM(edit_count) AS edits FROM fact_session_files WHERE file_path = '${api.topFiles[0].file_path.replace(/'/g, "''")}'`);
      if (compareValues('topFiles[0]_edits', api.topFiles[0].edits, firstFileResult[0]?.edits)) matches++;
      else mismatches++;
    }
    console.log();
  }

  // Top People
  if (api.topPeople && api.topPeople.length > 0) {
    console.log('=== Top People (sample) ===');
    if (api.topPeople[0]) {
      const firstPersonResult = await runQuery(`SELECT SUM(total_cost) AS cost FROM dim_sessions WHERE person_id = '${api.topPeople[0].person_id}'`);
      if (compareValues('topPeople[0]_cost', api.topPeople[0].total_cost, firstPersonResult[0]?.cost)) matches++;
      else mismatches++;
    }
    console.log();
  }

  const total = matches + mismatches;
  console.log(`\nOVERALL: ${mismatches === 0 ? 'ALL MATCH' : `${mismatches} MISMATCH(ES) / ${total} total`}`);

  conn.close();
  db.close();

  process.exit(mismatches > 0 ? 1 : 0);
}

main().catch((err) => {
  console.error('Fatal error:', err.message);
  process.exit(1);
});

File: verify_numbers.py
"""Reconciliation: compare /api/dashboard against DuckDB direct queries.

Run inside the watcher container (has python+duckdb+/data mount,
can reach frontend at http://frontend:3000 via the compose network).
"""
import json
import sys
import urllib.request

import duckdb

DB_PATH = "/data/read/aura.duckdb"
API_URL = "http://frontend:3000/api/dashboard?range=all"
TOL_ABS = 0.01
TOL_REL = 0.01


def fetch_api():
    with urllib.request.urlopen(API_URL, timeout=15) as r:
        return json.loads(r.read().decode("utf-8"))


def close(a, b):
    if a is None and b is None:
        return True
    if a is None or b is None:
        return False
    try:
        af = float(a)
        bf = float(b)
    except (TypeError, ValueError):
        return a == b
    if abs(af - bf) <= TOL_ABS:
        return True
    base = max(abs(af), abs(bf), 1.0)
    return abs(af - bf) / base <= TOL_REL


def show(label, api_v, db_v):
    ok = close(api_v, db_v)
    mark = "MATCH" if ok else "MISMATCH"
    print(f"{label:<40} api={api_v!s:<25} db={db_v!s:<25} {mark}")
    return ok


def one(con, sql):
    row = con.sql(sql).fetchone()
    return row[0] if row else None


def main():
    api = fetch_api()
    if not api or not api.get("kpis"):
        print("API returned empty payload — cannot reconcile.")
        sys.exit(2)

    con = duckdb.connect(DB_PATH, read_only=True)
    fails = 0

    print("=== KPIs ===")
    kp = api["kpis"]
    checks = [
        ("kpis.total_sessions",  kp.get("total_sessions"),
            one(con, "select count(distinct session_id) from dim_sessions")),
        ("kpis.total_cost",      kp.get("total_cost"),
            one(con, "select coalesce(sum(total_cost), 0) from dim_sessions")),
        ("kpis.total_turns",     kp.get("total_turns"),
            one(con, "select coalesce(sum(turn_count), 0) from dim_sessions")),
        ("kpis.total_apps",      kp.get("total_apps"),
            one(con, "select count(distinct cwd) from dim_sessions")),
        ("kpis.total_people",    kp.get("total_people"),
            one(con, "select count(distinct person_id) from dim_sessions where person_id is not null")),
        ("kpis.total_tool_calls", kp.get("total_tool_calls"),
            one(con, "select sum(tools_used) from dim_sessions")),
    ]
    for label, a, b in checks:
        if not show(label, a, b):
            fails += 1

    print("\n=== Section sizes (row counts) ===")
    section_checks = [
        ("topApps.rows",    len(api.get("topApps", []))),
        ("topAgents.rows",  len(api.get("topAgents", []))),
        ("toolMix.rows",    len(api.get("toolMix", []))),
        ("providers.rows",  len(api.get("providers", []))),
        ("models.rows",     len(api.get("models", []))),
        ("topFiles.rows",   len(api.get("topFiles", []))),
        ("topPeople.rows",  len(api.get("topPeople", []))),
        ("dailySpend.rows", len(api.get("dailySpend", []))),
        ("recentErrors.rows", len(api.get("recentErrors", []))),
    ]
    for label, n in section_checks:
        print(f"{label:<40} api={n}")

    print("\n=== Top-of-list spot checks ===")
    top_apps = api.get("topApps", [])
    if top_apps:
        a0 = top_apps[0]
        db_top_app = con.sql(
            "select app_id, sum(total_cost) as c from dim_sessions "
            "where app_id is not null group by 1 order by c desc limit 1"
        ).fetchone()
        if db_top_app:
            if not show("topApps[0].app_id", a0.get("app_id"), db_top_app[0]):
                fails += 1
            if not show("topApps[0].cost",   a0.get("total_cost"), db_top_app[1]):
                fails += 1

    top_agents = api.get("topAgents", [])
    if top_agents:
        a0 = top_agents[0]
        db_top_agent = con.sql(
            "select ds.agent, al.app_id, al.project_id, sum(ds.total_cost) as c "
            "from dim_sessions ds "
            "left join int_app_cwd_lookup al on al.cwd = ds.cwd and al.tenant_id = ds.tenant_id "
            "group by ds.agent, al.app_id, al.project_id "
            "order by c desc nulls last limit 1"
        ).fetchone()
        if db_top_agent:
            if not show("topAgents[0].agent", a0.get("agent"), db_top_agent[0]):
                fails += 1
            if not show("topAgents[0].cost",  a0.get("total_cost"), db_top_agent[3]):
                fails += 1

    daily = api.get("dailySpend", [])
    if daily:
        api_total = sum(float(d.get("cost") or 0) for d in daily)
        db_total = one(con, "select coalesce(sum(cost),0) from ("
                            "  select sum(daily_cost) as cost from fact_daily_spend "
                            "  group by date order by date desc limit 14"
                            ")")
        if not show("dailySpend.sum(cost) (14d)", api_total, db_total):
            fails += 1

    print()
    if fails == 0:
        print("OVERALL: ALL MATCH")
        sys.exit(0)
    else:
        print(f"OVERALL: {fails} MISMATCH(ES)")
        sys.exit(1)


if __name__ == "__main__":
    main()
